nodes/generators_extended/torus2D.py

```diff
@@ -0,0 +1,256 @@
+        default=0.0, min=0.0, soft_min=0.0,
+        update=updateNode)
+
+    # OTHER options
+    Separate = BoolProperty(
+        name='Separate', description='Separate UV coords',
+        default=False,
+        update=updateNode)
+
+    def sv_init(self, context):
+        self.width = 180
+        self.inputs.new('StringsSocket', "R").prop_name = 'torus_R'
+        self.inputs.new('StringsSocket', "r").prop_name = 'torus_r'
+        self.inputs.new('StringsSocket', "n1").prop_name = 'torus_n1'
+        self.inputs.new('StringsSocket', "n2").prop_name = 'torus_n2'
+        self.inputs.new('StringsSocket', "rP").prop_name = 'torus_rP'
+
+        self.outputs.new('VerticesSocket', "Vertices")
+        self.outputs.new('StringsSocket',  "Edges")
+        self.outputs.new('StringsSocket',  "Polygons")
+
+    def draw_buttons(self, context, layout):
+        layout.prop(self, "Separate", text="Separate")
+        layout.prop(self, 'mode', expand=True)
+
+    def process(self):
+        # return if no outputs are connected
+        if not any(s.is_linked for s in self.outputs):
+            return
+
+        # input values lists (single or multi value)
+        # list of MAJOR or EXTERIOR radii
+        input_RR = self.inputs["R"].sv_get()[0]
+        # list of MINOR or INTERIOR radii
+        input_rr = self.inputs["r"].sv_get()[0]
+        # list of number of MAJOR sections : RADIAL
+        input_n1 = self.inputs["n1"].sv_get()[0]
+        # list of number of MINOR sections : CIRCULAR
+        input_n2 = self.inputs["n2"].sv_get()[0]
+        # list of RADIAL phases
+        input_rp = self.inputs["rP"].sv_get()[0]
+
+        # sanitize the input values
+        input_RR = list(map(lambda x: max(0, x), input_RR))
+        input_rr = list(map(lambda x: max(0, x), input_rr))
+        input_n1 = list(map(lambda x: max(3, int(x)), input_n1))
+        input_n2 = list(map(lambda x: max(2, int(x)), input_n2))
+
+        # convert input radii values to MAJOR/MINOR, based on selected mode
+        if self.mode == 'EXT_INT':
+            # convert radii from EXTERIOR/INTERIOR to MAJOR/MINOR
+            # (extend radii lists to a matching length before conversion)
+            input_RR, input_rr = match_long_repeat([input_RR, input_rr])
+            input_R = list(map(lambda x, y: (x + y) * 0.5, input_RR, input_rr))
+            input_r = list(map(lambda x, y: (x - y) * 0.5, input_RR, input_rr))
+        else:  # values already given as MAJOR/MINOR radii
+            input_R = input_RR
+            input_r = input_rr
+
+        parameters = match_long_repeat([input_R, input_r, input_n1, input_n2, input_rp])
+
+        V, E, P = [self.outputs[n] for n in ['Vertices', 'Edges', 'Polygons']]
+        for s, f in [(V, torus_verts), (E, torus_edges), (P, torus_polygons)]:
+            if s.is_linked:
+                s.sv_set([f(n1, n2) if s != V else f(R, r, n1, n2, p, self.Separate)
+                          for R, r, n1, n2, p in zip(*parameters)])
```

Comment:
i do like iterating over VEP like that. But linesplitting the list comprehension is murky territory (especially with a nested if/else), It's a sign we're forcing things. We can still be explicit without being verbose .

I favour less nested over vertical compression :)

```python
        V, E, P = self.outputs[:]
        params = match_long_repeat([input_R, input_r, input_n1, input_n2, input_rp])

        for s, f in [(V, torus_verts), (E, torus_edges), (P, torus_polygons)]:
            if not s.is_linked:
                continue
            if s == V:
                s.sv_set([f(self.Separate, *args) for args in zip(*params)])
                # s.sv_set([f(R, r, n1, n2, p, self.Separate) for R, r, n1, n2, p in zip(*params)])
            else:
                s.sv_set([f(n1, n2) for _, _, n1, n2, _ in zip(*params)])
```